### ISSUE TITLE
No UDP via 3proxy SOCKS5 inside Docker

### ISSUE THREAD

**Issue by ducmthai** [z3APA3A/3proxy#565]
Hi,

It seems that 3proxy inside a docker container cannot proxy UDP over SOCKS5. I tested 3proxy both on the host machine and in a container. Only on the host I can connect to the UDP tracker in qBittorrent. Here are my setups and results:

# On the host
qBittorrent client worked fine. There are a number of DHT nodes. Tested with doileak.com's torrent leak test. My IP appeared on doileak torrent test result.

## 3proxy.cfg:

```Properties
writable
nserver 1.1.1.1
nserver 1.0.0.1
nscache 65536
log
logformat "L%t.%.  Err:%E Client:%C:%c Remote:%R:%r ExternalIP:%e:%q ReqRecv:%O:%I %h %T"
monitor ./3proxy.cfg
admin -p8080
socks -Dieth0 -p2128
```

## 3proxy log
```
3proxy ./3proxy.cfg       
1606959865.952  Err:00000 Client:0.0.0.0:8080 Remote:0.0.0.0:0 ExternalIP:0.0.0.0:0 ReqRecv:0:0 0 Accepting connections [3569450/2465509856]
1606959865.953  Err:00000 Client:0.0.0.0:2128 Remote:0.0.0.0:0 ExternalIP:0.0.0.0:0 ReqRecv:0:0 0 Accepting connections [3569450/2457117152]
                                                                               
1606959983.277  Err:00000 Client:192.168.0.28:63423 Remote:<my_external_ip>:36510 ExternalIP:192.168.0.32:36510 ReqRecv:0:0 0 CONNECT <my_external_ip>:36510
1606959991.281  Err:00013 Client:192.168.0.28:63422 Remote:114.145.85.124:43563 ExternalIP:0.0.0.0:43563 ReqRecv:0:0 0 CONNECT 114.145.85.124:43563
1606960043.368  Err:00000 Client:192.168.0.28:63424 Remote:<my_external_ip>:36510 ExternalIP:192.168.0.32:36510 ReqRecv:0:0 0 CONNECT <my_external_ip>:36510
1606960044.370  Err:00000 Client:192.168.0.28:63425 Remote:<my_external_ip>:36510 ExternalIP:192.168.0.32:36510 ReqRecv:0:0 0 CONNECT <my_external_ip>:36510
```
# In a container
Same qBittorent client did not work. There was no DHT node. doileak test showed no IP.

## Container route table:

```Shell
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
default         10.8.3.1        128.0.0.0       UG    0      0        0 tun0
default         10.29.0.1       0.0.0.0         UG    0      0        0 eth0
10.8.3.0        *               255.255.255.0   U     0      0        0 tun0
10.29.0.0       *               255.255.255.0   U     0      0        0 eth0
96.9.243.171    10.29.0.1       255.255.255.255 UGH   0      0        0 eth0
128.0.0.0       10.8.3.1        128.0.0.0       UG    0      0        0 tun0
192.168.0.0     10.29.0.1       255.255.255.0   UG    0      0        0 eth0
```

## 3proxy.cfg:

```Properties
writable
nserver 1.1.1.1
nserver 1.0.0.1
nscache 65536
counter "/etc/3proxy/3proxy.3cf" D "/etc/3proxy/traf"
log
logformat "L%t.%.  Err:%E Client:%C:%c Remote:%R:%r ExternalIP:%e:%q ReqRecv:%O:%I %h %T"
monitor /etc/3proxy/3proxy.cfg
admin -p8080
# with or without -Detun0 it's the same
socks -Dieth0 -p2128 -Detun0
```

## curl test from 192.168.0.28
```
$ curl -x socks5h://192.168.0.32:2128 -L ifconfig.co/json
{
  "ip": "96.9.243.172",
  ...
  }
}%
```

## 3proxy log:
```Shell
1606960324.436  Err:00000 Client:0.0.0.0:8080 Remote:0.0.0.0:0 ExternalIP:0.0.0.0:0 ReqRecv:0:0 0 Accepting connections [293/2798480136]
1606960324.436  Err:00000 Client:0.0.0.0:2128 Remote:0.0.0.0:0 ExternalIP:0.0.0.0:0 ReqRecv:0:0 0 Accepting connections [293/2795182856]
1606960358.257 Err:00000 Client:192.168.0.28:63524 Remote:104.28.18.94:80 ExternalIP:10.8.3.2:80 ReqRecv:79:1297 0 CONNECT ifconfig.co:80

# Entry when exit qBittorrent
1606960667.179  Err:00000 Client:192.168.0.28:63469 Remote:0.0.0.0:0 ExternalIP:0.0.0.0:0 ReqRecv:0:0 0 UDPMAP 0.0.0.0:0
```

Did I missed anything in the configuration? Thank you for any help you can offer!

**Comment by 215php** [comment]
请问3proxy怎么代理udp?

**Comment by z3APA3A** [comment]
SOCKSv5 opens  2 additional UDP ports for every UDP mapping (one for traffic exchange with client and one with server), so you need to (somehow) map all ephemeral UDP ports to container. I'm not docker guru, but I think it can be achieved with `--network host`

**Comment by ducmthai** [comment]
Thanks for the tip. I've spent hours trying different setups without realizing the host mode would work. You made my day!
The sad part is I'm running another setup inside WSL2 and host mode is not supported yet.

**Comment by z3APA3A** [comment]
I think you can try to add an additional IP address to host and map all ports for this IP to container with -p IP:1025-65535:1025-65535 (you may need to use -N option to proxy to specify the NAT IP).

**Comment by ducmthai** [comment]
Exposing that number of ports would take time for the container to start up since Docker will do iptables stuff underneath. On the other hand, host mode network makes tun0 as default on the host. I have a few other VPN containers on the same host so this is not ideal. I switched to a dedicated lightweight VM instead. 

I also noticed without -s0 in socks, I couldn't upload any photos to google photos. It would be useful to put this note for folks running 3proxy on WSL2.

**Comment by z3APA3A** [comment]
What is 3proxy version you are experience problems without -s0 and do you have same problems with VM?

**Comment by ducmthai** [comment]
It's 0.9.3. Initially, I ran 3proxy without -s0 in both VM and WSL2 but only noticed the problem with WSL2. I moved away from WSL2 in one of my setups but I thought it's better to have -s0 everywhere :) Will that cause any potential issues?

**Comment by z3APA3A** [comment]
I'm interested in fixing the issue, if there is any.

-s0 disables splice, splice reduces CPU usage but incereases IOPS. Under older WSL (without docker) i was able to get 10x higher throughput via proxy with splice enabled between windows client and linux server service. 

I tested with both FF and Chrome with google disk through socks running in docker in WSL2 and I see no issues specific to -s. I had some issues with VPN connection, which looked correlated with splice usage, but it can not be directly linked to proxy (I was uploading photos from network disk attached via VPN, but traffic to Google though proxy was not routed to VPN). May be it's some glitch in Windows, WSL or docker due to higher IOPS. With docker on WSL2 I see -s0 takes ~2x more CPU (docker uses more CPU than Chrome with -s0 and significantly less CPU than Chrome without -s0 during upload if uploading from local disk on 100Mbps connection).

**Comment by ducmthai** [comment]
My test was done with both FF and Chrome, both through FF native proxy config and Proxifier v4. The traffic to Google was routed through 3proxy. I did not monitor CPU usage at the time but I felt that downlink is quite good but the uplink is size-capped at 1MB. In 3proxy, the error code for such connections (photos.google.com:443) is 0094. In Proxifier, the connection sent 1MB before it's closed. WSL2 uses Microsoft kernel so I think it's the curse. I added the -s0 after reading about optimization on the Document page.
